### PR TITLE
Add nested mlflow option to enable_mlflow decorator

### PR DESCRIPTION
**Pull request by Val3nt-ML**
## Describe changes
We add a parameter `nested` to enable_mlflow decorator to create nested run for each step of a Zen-ML pipeline in MLFlow.

_Example_:
Here is the syntax to use it : 
```python
@enable_mlflow(nested=True)
@step(enable_cache=True)
def my_func(df: pd.DataFrame) -> Output(output_df=pd.DataFrame):
    outpuf_df = my_process(df)
    mlflow.log_param("my_param", 42)
    return output_df
```

And here is the render in MlFlow :
image

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by htahir1** [comment]
@Val3nt-ML  Thank you for the great contribution! We'll leave reviews soon but would be great if you could sign the CLA as indicated by the bot above :-)

**Comment by Val3nt-ML** [comment]
Hello Hamza,
I will fix it :)

Kind regards
Valentin LAURENT

________________________________
De : Hamza Tahir ***@***.***>
Envoyé : Wednesday, June 29, 2022 7:11:44 PM
À : zenml-io/zenml ***@***.***>
Cc : Valentin Laurent ***@***.***>; Mention ***@***.***>
Objet : Re: [zenml-io/zenml] [WIP] Feat/add nested mlflow (PR #742)


@Val3nt-ML<https://github.com/Val3nt-ML> Thank you for the great contribution! We'll leave reviews soon but would be great if you could sign the CLA as indicated by the bot above :-)

**Comment by htahir1** [comment]
Thank you! Im also a bit unsure about this nesting from a user perspective. So if I set nesting as `true` then what happens? Does every `log_param` call make a new run? What are the benefits from a user perspective? We should add these to the docs/examples somewhere (Perhaps in the mlflow_tracking example?)

**Comment by lopezco** [comment]
> Thank you! Im also a bit unsure about this nesting from a user perspective. So if I set nesting as `true` then what happens? Does every `log_param` call make a new run? What are the benefits from a user perspective? We should add these to the docs/examples somewhere (Perhaps in the mlflow_tracking example?)

If `nested=True` all the mlflow log functions inside the step will log to the nested run. In contrast, if  `nested=False`, the logs will go to the main active run.

**Comment by lopezco** [comment]
@htahir1 the benefits will be that you can log different metrics or parameters on some steps in separated runs that will be nested into the main pipeline run. 

An example on how we are currently using this is to track the parameters used in some steps and how many rows we are dropping in the filter steps as a metric

**Comment by htahir1** [comment]
@lopezco Thank you so much for the clarification. Super useful feature! I think we might need some visibility on this. I can see two places where it would be awesome to add in this PR:

- Updating our [example for mlflow tracking](https://github.com/zenml-io/zenml/tree/main/examples/mlflow_tracking) to reflect this. Perhaps we can add a `nested` section at the bottom with a seperate note explaining that behavior.
- Adding this in the actual code in the docstring of the decorator

WDYT?

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 1]
```diff
@@ -51,6 +51,7 @@ def enable_mlflow() -> Callable[[S], S]:
 
 def enable_mlflow(
     _step: Optional[S] = None,
+    nested: Optional[bool] = False
```
[the remark this answers is not in the document]
Comment (lopezco) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py:54]: In the docstrings or in the example folder?

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 2]
```diff
@@ -95,9 +96,10 @@ def entrypoint(
 
     Args:
         _step: The decorated step class.
+        nested: Enable to have nested run during mlflow logging
```
[the remark this answers is not in the document]
Comment (lopezco) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py]: What do you think about the new docstring including an example?

**Comment by htahir1** [comment]
Thank you! Looks really good. I just see some small problems now with linting and formatting. Would you be so kind as to run the following scripts to ensure coding guideliens are met?

```
poetry run bash scripts/format.sh
poetry run bash scripts/lint.sh
poetry run bash scripts/check-spelling.sh
```

This should take care of the issue as it seems to just be superfluous or missing whitespaces.

**Review by schustmi — CHANGES_REQUESTED** [review]
Just two small comments on type annotations, otherwise looks great!

**Review by schustmi — APPROVED** [review]
LGTM!

**Comment by htahir1** [comment]
Awesome, I will merge this now after the tests pass. Thank you so much for the contribution to the folks at Riverbank!

**Pull request merged**